### PR TITLE
feat: make federated modules dynamic

### PR DESCRIPTION
**Pull request by riccardo-forina**

### CODE REVIEW

**Comment by riccardo-forina** [comment]
This is a PoC of dynamic federated modules. Right now it's an almost direct ripoff of [this example](https://github.com/module-federation/module-federation-examples/blob/master/dynamic-system-host/app1/src/App.js), with a twist: instead of hardcoding the remotes in the code, we load them from a single source of truth.

The idea is for the host app to load once at boot time a json containing all the allowed federated modules, as a dictionary of `scope-id: remote-url` (eg. `mkUiFrontend: https://localhost:9000`. In the code we then can load the federated modules asynchronously providing just the scope id, the module we want to be loaded, and a render prop to actually use the loaded module:

```jsx
{/* somewhere in the root of the host app */}
<FederatedModuleProvider config="/application-services/openshift-streams/federated-modules.json">
	...
	{/* where we need to load a federated module */}
	<FederatedModule
	    scope="mkUiFrontend"
	    module="./OpenshiftStreams"
	    render={
			OpenshiftStreamsFederated => 
				{/* we can pass the required props to the loaded module */}
				<OpenshiftStreamsFederated token={token} /> 
		}
	/>
	...
</FederatedModuleProvider>
```

One thing to consider is that with dynamic federated modules we need to use default exports because `React.lazy` [requires it](https://reactjs.org/docs/code-splitting.html#reactlazy): `React.lazy takes a function that must call a dynamic import(). This must return a Promise which resolves to a module with a default export containing a React component.`

If we like this approach, I'll update this PR to:

* move the components and functions in their own package. This will allow nested dynamic federated modules while keeping a single source of truth, which should simplify the ops side of the problem.
* use Typescript and make the `FederatedModule` generic. This will give us typechecking at build time using a syntax like this:
```tsx
import { OpenshiftStreamsFederatedProps } from "mkUiFrontend/OpenshiftStreams";

<FederatedModule<OpenshiftStreamsFederatedProps>
    scope="mkUiFrontend"
    module="./OpenshiftStreams"
    render={
		OpenshiftStreamsFederated => 
			{/* this line will give a compile time error because of the missing `token` prop */}
			<OpenshiftStreamsFederated /> 
	}
/>
```
* handle loading errors and edge cases
* add tests

CC @karelhala @redallen @Hyperkid123 @pmuir

**Comment by Hyperkid123** [comment]
@riccardo-forina This seems like a good start.

We are working on something similar. If you are interested take a look at this: https://github.com/scalprum/scaffloding/blob/main/packages/react-core/src/scalprum-component.tsx.

In summary, it works like this. Each app (Federated module) generates a manifest file at build time with entry points. We do this because we are using hashes even for container files in order to bypass cache and we can't statically define the container remote. The async component loads the manifest on the mount and then registers the remote chunk to webpack at runtime. Rest is just the classic react lazy -> suspense interaction.

Currently, it requires additional dependencies which will cause conflicts with insights chrome 2, but we plan to generalize it and provide the API to chrome apps so they can use the same mechanism. I can't tell you when exactly this will be available but we plan to release the insights chrome 2.0 to the beta environment in early 2021.

**Comment by riccardo-forina** [comment]
Thanks for this, it looks very well implemented. From what I see modules are loaded immediately after parsing the manifest, am I understanding this right? It's a different take on the problem, we want to load modules only when strictly necessary and not have to rebuild the app if something changes on the remote side, be that a change of url or version.

**Comment by Hyperkid123** [comment]
@riccardo-forina yes the module is loaded and registered immediately. We will control the "when" via a router, but any React.lazy wrapper around this component will achieve the same. We want to make this component generic and let the consumer control when the remote is registered.

We also won't be rebuilding the app when remote changes, that is why we pull the manifest from CDN. So basically you would always get the "fresh" remote when you use this component.

**Comment by karelhala** [comment]
@riccardo-forina not really after parsing the manifest but once the route renders, we register the manifest called `fed-mods` (you can name it whatever you want, it just has to be accessible on same origin) under certain app name, once the user requires to load the app (route change) we call an async function to fetch requested module from specified path.

On top of this each app can define multiple modules, meaning they can have one entry point (this is required to render the application) and multiple federated modules to share between different apps.

### Manifest structure
```JS
{
  [appName]: {
    entry: [ `${fullPath}.js` ]
    modules: []
  }
}
```
* `appName` - each app can define more than one federated module (don't mix it with `scope`), you can achieve this by having multiple `ModuleFederationPlugin` in `plugins` field of your webpack config or by simply combining multiple config builds
* `entry` - each record can have multiple entry files, one app could consist of more than one federated module
* `modules` - this is just dependendecies for each entry. It's currently no in use, but should the need for side loading these modules arise we can consume them

### Usage

The flow of creating creating new federated component is:
* Create manifest
* Define module federation
* Update the list of available modules
* Use async component when needed

#### Create manifest
Each entry then consits of classic federated module. So let's take an example of manifest located on url `/apps/someApp/fed-mods.json`

```JS
{
  app1: {
    entry: ['/apps/someApp/app.f44r56r67r78r881.js']
  },
  app2: {
    entry: ['/apps/someApp/app.rew44t8rt4t48t22.js']
  }
}
```

#### Define module federation
These files `'/apps/someApp/app.f44r56r67r78r881.js'` and `'/apps/someApp/app.rew44t8rt4t48t22.js'` were created in webpack as:

```JS
{
  //.. rest of webpack config
  plugins: [
    //.. rest of webpack plugin's config
    new ModuleFederationPlugin({
      name: 'app1',
      filename: 'app.[hash].js',
      exposes: {
        './Widget1': resolve(__dirname, '../src/js/someFile.js'),  // ref: app1/Widget1
        './Widget2': resolve(__dirname, '../src/js/someFile2.js') // ref: app1/Widget2
      },
      shared: [
        { react: { singleton: true, requiredVersion: deps.react } },
        { 'react-dom': { singleton: true, requiredVersion: deps['react-dom'] } },
        { 'react-router-dom': { singleton: true, requiredVersion: deps['react-router-dom'] } },
        { '@patternfly/react-table': { singleton: true } },
        { '@patternfly/react-core': { singleton: true } },
        { 'react-redux': { singleton: true } },
    ],
    new ModuleFederationPlugin({
      name: 'app2',
      filename: 'app.[hash].js',
      exposes: {
        './Widget1': resolve(__dirname, '../src/js/anotherFile.js'), // ref: app2/Widget1
        './Widget2': resolve(__dirname, '../src/js/differentFile.js') // ref: app2/Widget2
      },
      shared: [
        { react: { singleton: true, requiredVersion: deps.react } },
        { 'react-dom': { singleton: true, requiredVersion: deps['react-dom'] } },
        { 'react-router-dom': { singleton: true, requiredVersion: deps['react-router-dom'] } },
        { '@patternfly/react-table': { singleton: true } },
        { '@patternfly/react-core': { singleton: true } },
        { 'react-redux': { singleton: true } },
    ]
  })
  ]
}
```
#### Update the list of available modules
We will register this config via (done in insights-chrome)
```JS
useScalprum({
  someApp: {
    appName: 'someApp',
    manifestLocation: `${window.location.origin}/apps/someApp/fed-mods.json`,
  }
})
```
#### Use async component when needed
You can write your own module accessor, or use `ScalprumComponent` (recommended), this wil use `ref: app1/Widget1`
```JS
<ScalprumComponent appName="someApp" module="./Widget1" scope="app1" />
```

File: src/Components/FederatedModule.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        return null;
+    }
+
+    const Component = React.lazy(
+        loadComponent(scope, module)
+    );
+
+    return (
+        <React.Suspense fallback="Loading System">
+            {render(Component)}
```
Comment (karelhala) [inline src/Components/FederatedModule.js:116]: Interesting take on this, if you don't want to require the `render` function you can also change this to
```JSX
export function FederatedModule({ scope, module, render, ...props }) {
    const url = useFederatedModule(scope);
    const { ready, failed } = useDynamicScript({ url });

    if (!ready || failed) {
        return null;
    }

    const Component = React.lazy(
        loadComponent(scope, module)
    );

    return (
        <React.Suspense fallback="Loading System">
            <Component {...props} />
        </React.Suspense>
    );
}

FederatedModule.propTypes = {
    scope: PropTypes.string,
    module: PropTypes.string
};
```

Usage then changes to
```JSX
<FederatedModule
  scope="mkUiFrontend"
  module="./OpenshiftStreams"
  token={token}
/>
```

**Comment by karelhala** [comment]
> One thing to consider is that with dynamic federated modules we need to use default exports because React.lazy

If you have named export you can use it as
```JS
const Component = React.lazy(() => import('./someCmp').then(data => ({ default: data.NamedExport })))
```

**Comment by Hyperkid123** [comment]
> > One thing to consider is that with dynamic federated modules we need to use default exports because React.lazy
> 
> If you have named export you can use it as
> 
> ```js
> const Component = React.lazy(() => import('./someCmp').then(data => ({ default: data.NamedExport })))
> ```

I would not play with fire here. If you have multiple named exports and use dynamic import, you will pull all the code at once, even though you use only one component. This is not "treeshakeable" so you waste a lot of bandwidth.

Plus if you have multiple components pointing to the same dynamic import, you will always pull the shole file so the efficiency of lazy-loaded chunks goes out of the window. instead of pulling one component only once you will pull ALL the components from the file X times.

Also if you split the file you give webpack a chance to create common chunks which may further decrease the total size of async modules.

**Comment by karelhala** [comment]
@Hyperkid123 yeah sure, default exports are preferable, but if you have no other choice there's always a this solution.

**Comment by riccardo-forina** [comment]
Thanks for your feedback, we will keep using our solution but not develop it further while we wait for the release of this functionality as part of Insights Chrome 2. Please keep us posted for updates!
